src/wp-includes/meta.php

```diff
@@ -958,30 +960,220 @@ function sanitize_meta( $meta_key, $meta_value, $meta_type ) {
 	 * @param string $meta_type  Meta type.
 	 */
-	return apply_filters( "sanitize_{$meta_type}_meta_{$meta_key}", $meta_value, $meta_key, $meta_type );
+	$meta_value = apply_filters( "sanitize_{$object_type}_meta_{$meta_key}", $meta_value, $meta_key, $object_type );
+
+	/**
+	 * Filter the sanitization of a specific meta key of a specific meta object type and subtype.
+	 *
+	 * The dynamic portions of the hook name, `$object_type`, `$object_subtype`, and `$meta_key`
+	 * refer to the metadata object type (comment, post, or user), the subtype of that object type
+	 * and the meta key value respectively.
+	 *
+	 * @since 4.5.0
+	 *
+	 * @param mixed  $meta_value     Meta value to sanitize.
+	 * @param string $meta_key       Meta key.
+	 * @param string $object_type    Meta type.
+	 * @param string $object_subtype Subtype of the object type.
+	 */
+	return apply_filters( "sanitize_{$object_type}_{$object_subtype}_meta_{$meta_key}", $meta_value, $meta_key, $object_type, $object_subtype );
 }
 
 /**
- * Register meta key
+ * Register a meta key.
+ *
+ * @todo should we wrap `register_meta()` with `register_meta_key()` to be more explicit?
  *
  * @since 3.3.0
+ * @since 4.6.0 Modified to support an array of data to attach to registered meta keys. Previous arguments for
+ *              `$sanitize_callback` and `$auth_callback` have been folded into this array.
+ *
+ * @param string $object_type    Type of object this meta is registered to.
+ * @param string $meta_key       Meta key to register.
+ * @param array  $args {
+ *     Data used to describe the meta key when registered.
  *
- * @param string       $meta_type         Type of meta
- * @param string       $meta_key          Meta key
- * @param string|array $sanitize_callback A function or method to call when sanitizing the value of $meta_key.
- * @param string|array $auth_callback     Optional. A function or method to call when performing edit_post_meta, add_post_meta, and delete_post_meta capability checks.
+ *     @type string   $sanitize_callback A function or method to call when sanitizing `$meta_key` data.
+ *     @type string   $auth_callback     Optional. A function or method to call when performing edit_post_meta,
+ *                                       add_post_meta, and delete_post_meta capability checks.
+ *     @type bool     $public            Whether data associated with this meta key can be considered public.
+ *     @type array    $schema            The meta field's schema, according to JSON schema.
+ * }
+ * @return true|WP_Error
  */
-function register_meta( $meta_type, $meta_key, $sanitize_callback, $auth_callback = null ) {
-	if ( is_callable( $sanitize_callback ) )
-		add_filter( "sanitize_{$meta_type}_meta_{$meta_key}", $sanitize_callback, 10, 3 );
+function register_meta( $object_type, $meta_key, $args ) {
+	global $wp_meta_manager;
+	// @todo validate object type as post, comment, user, term?
+
+	// @todo should this be an object similar to post types?
+	$args = array(
```

Comment:
If $args is defined properly in a call to `register_meta( $object_type, $meta_key, $args )` then that never gets used, we should fix that or use `$default` as the array here, merging it with $args after the `if ( is_callable( $passed_args[0] ) )` callable check.